From eb0a7cebf1d28325939d8a4eff18c1b629c4480c Mon Sep 17 00:00:00 2001
From: Dongwon Kim <dongwon.kim@intel.com>
Date: Mon, 4 Jan 2021 21:26:25 -0800
Subject: [PATCH 2/2] gl-shader: convert to linear space and back to srgb only
 when needed

Converting color space to linear is only needed when color conversion
and/or HDR/SDR tone mapping is required (e.g. HDR output). If none of
these conversions/adjustments are needed, we can skip srgb->linear and
linear->srgb processes that consumes lots of GPU cycles.

Signed-off-by: Dongwon Kim <dongwon.kim@intel.com>
---
 libweston/renderer-gl/gl-shaders.c | 19 +++++++++++++------
 1 file changed, 13 insertions(+), 6 deletions(-)

diff --git a/libweston/renderer-gl/gl-shaders.c b/libweston/renderer-gl/gl-shaders.c
index 715f9677..03236b17 100644
--- a/libweston/renderer-gl/gl-shaders.c
+++ b/libweston/renderer-gl/gl-shaders.c
@@ -478,22 +478,29 @@ generate_hdr_process_shader(struct gl_shader_source *shader_source,
 		(requirements->tone_mapping == SHADER_TONE_MAP_HDR_TO_HDR) ||
 		(requirements->tone_mapping == SHADER_TONE_MAP_SDR_TO_HDR);
 
-	if (requirements->degamma)
+	uint32_t need_linear_conversion = requirements->csc_matrix |
+					  requirements->tone_mapping;
+
+	if (need_linear_conversion && requirements->degamma)
 		gl_shader_source_add(shader_source, eotf_shader);
 
-	if (requirements->csc_matrix)
+	if (requirements->csc_matrix) {
 		gl_shader_source_add(shader_source, csc_shader);
+	}
 
-	if (requirements->degamma && need_range_increment)
+	if (requirements->degamma && need_range_increment) {
 		gl_shader_source_add(shader_source, sl_shader);
+	}
 
-	if (requirements->tone_mapping)
+	if (requirements->tone_mapping) {
 		gl_shader_source_add(shader_source, hdr_shader);
+	}
 
-	if (requirements->nl_variant)
+	if (need_linear_conversion && requirements->nl_variant) {
 		gl_shader_source_add(shader_source, nl_shader);
+	}
 
-	if (requirements->gamma)
+	if (need_linear_conversion && requirements->gamma)
 		gl_shader_source_add(shader_source, oetf_shader);
 
 }
-- 
2.17.1

